src/approx/terrain.py

```diff
@@ -245,195 +252,184 @@ def build_ls_matrix(u_knots, v_knots, terrain, sparse=False):
+            # possible problem: u x v
+            mat_b[idx] = numpy.kron(u_base_vec, v_base_vec)
+
         interval[idx][0] = i_idx
         interval[idx][1] = j_idx
 
-    return mat_b, interval
+    if sparse is True:
+        mat_b = scipy.sparse.csr_matrix((data,(row,col)),shape=(terrain_len, u_n_basf * v_n_basf))
 
+    return mat_b, interval
 
-def build_reg_matrix(u_knots, v_knots, sparse=False):
+def build_reg_matrix(u_knots, v_knots, sparse):
     """
-    Regularization matrix for minimizing variation of the terrain surface
+    Construction of the regularization matrix (A) to decrease variation of the terrain
+    B z = b  ---> (B^T B + A)z = B^T b
     :param u_knots:
     :param v_knots:
     :param sparse:
     :return:
     """
+    import sys
+
     u_n_basf = len(u_knots) - 3
     v_n_basf = len(v_knots) - 3
     u_n_inter = len(u_knots) - 5
     v_n_inter = len(v_knots) - 5
     
-    q_points = [0, (1/2 - 1/math.sqrt(20)), (1/2 + 1/math.sqrt(20)), 1]
-    weights = [1/6, 5/6, 5/6, 1/6]
+    q_points = [0, (0.5 - 1/math.sqrt(20)), (0.5 + 1/math.sqrt(20)), 1]
+    weights = [1.0/6, 5.0/6, 5.0/6, 1.0/6]
     n_points = len(q_points)
 
-    u_point_val = numpy.empty((u_n_basf, u_n_inter * n_points))
-    ud_point_val = numpy.empty((u_n_basf, u_n_inter * n_points))
-
-    n = 0
-    for i in range(0, u_n_inter-1):
-        us = u_knots[i+2]
-        uil = u_knots[i+3] - u_knots[i+2]
-        print(n_points)
-        my_list = range(0, n_points-1)
-        print(my_list)
-        for j in my_list:
-            up = us + uil * q_points[j]
-            u_point_val[:, n] = spline_base_vec(u_knots, up)
-            ud_point_val[:, n] = spline_base_vec_diff(u_knots, up)
-            n += 1
-
-    v_point_val = numpy.empty((v_n_basf, v_n_inter * n_points))
-    vd_point_val = numpy.empty((v_n_basf, v_n_inter * n_points))
-
-    n = 0
-    for i in range(0, v_n_inter-1):
-        vs = v_knots[i+2]
-        vil = v_knots[i+3] - v_knots[i+2]
-        print(n_points)
-        my_list = range(0, n_points-1)
-        for k in my_list:
-            vp = vs + vil * q_points[k]
-            v_point_val[:, n] = spline_base_vec(v_knots, vp)
-            vd_point_val[:, n] = spline_base_vec_diff(v_knots, vp)
-            n += 1
-
-    a_mat = 0.0
-    # Matrix construction
-    for i in range(0, v_n_inter-1):
-        for k in range(0, n_points-1):
-            v_point = v_point_val[:, i * n_points + k]
-            vd_point = vd_point_val[:, i * n_points + k]
-            for l in range(0, u_n_inter-1):
-                for m in range(0, n_points-1):
-                    u_point = v_point_val[:, l * n_points + m]
-                    ud_point = vd_point_val[:, l * n_points + m]
-                    vd = numpy.kron(vd_point, u_point)
-                    ud = numpy.kron(v_point, ud_point)
-                    a_mat += weights[m] * weights[k] * (numpy.kron(ud, ud.transpose()) + numpy.kron(vd, vd.transpose()))
-
-    return a_mat
-
-
```

Comment:
Why you deleted this method? Unit test do not pass then. Do you aware of unit tests?